### ISSUE TITLE
XEP: Embedding data-objects in STDOUT

### ISSUE THREAD

**Issue by traverseda** [xonsh/xonsh#2598]
It would be nice if commands that outputted json were automatically turned into python objects. It's not a big deal, but once you move on from just json it could let us create some *very* interesting functionality.

By assigning each dataObject "tag" a uuid, we deal with escaping problems. Apps don't need to worry about escaping special characters, as long as they don't accidentally include the parent tags uuid. A string like

    #! dataObject open 16723b1d-e0c2-4fd1-a588-909bc716d761 aplication/json

would preface embedded data objects. dataObject stream *shouldn't* need to close, as we know when the software outputting them closes.

Later on, we can add support for a safe pickle equivalent and eventually remote python objects, and the ability to hold references to objects running in separate processes.

---

What I would like to know is roughly where I'd hook into xonsh to provide stdout checking and eventually redirection, and I'd like to have a few other eyes on the design before I get too deep into it.

**Comment by scopatz** [comment]
@traverseda that would be an awesome feature to have! It is a little unclear on what exactly the interface could / should be.  Putting something in the shebang line is certainly one option, but a program may not always output data of the same type, which would be hard to implement in a shebang line alone.  

Another option would be that the program could output some hidden escape sequences that xonsh knows how to monitor for.  This would also allow adding the python object hooks to 3rd party programs by wrapping them in an alias that sets the escape.

The places to look for adding these hooks are the CommandPipeline and the SubprocSpec

**Comment by traverseda** [comment]
Thanks for the information. I'm not sure we want *hidden* escape sequences. Visible escape sequences should be a lot less confusing if your shell doesn't support data objects. I don't see why you couldn't have multiple shebang lines in the *output* of a single program.

**Comment by scopatz** [comment]
Hmmm I definitely don't think that this is somethings users will want to see, and if it is printined to stdout, it could mess with the interaction with other programs in a pipeline.  I guess this should only ever apply to the last command in a pipeline (even if early commands have to-Python-object convertors registered).

I also do think it should be an option for users to see what converter  is being used. I am totally in favor of that. But by default it should be turned off. 

Also note that shebang lines are an Unix/posix interface for how to run commands, not what to do with the command after it has been executed.

**Comment by scopatz** [comment]
Sorry for being short here, have to run to class!

**Comment by traverseda** [comment]
>Also note that shebang lines are an Unix/posix interface for how to run commands, not what to do with the command after it has been executed.

Yeah, I'm torn on having them pull double duty here. By ~using~ abusing shebang lines like this, we might make it easier to hack in support for other shells. But it is a somewhat confusing way of doing things, and my plans in that direction are still half-baked.

>it could mess with the interaction with other programs in a pipeline.

I really don't see any other way to mark up the data without doing that. Obviously we'd need to provide a filter command that filters out those marks. Or that adds them to untyped output. Personally, I think that's worth it.

Of course the shell can always hide that output, if it support data objects. I think it's pretty important that those tags are visible when you start piping around more esoteric stuff, like jpeg images. If you see a binary or base64 encoded jpeg in your terminal, it's important that you be able to see what the hell went wrong.

So I'd say that by default xonsh should be stripping those tags from command output, definitely. Just that they should be visible in the stdout stream.

**Comment by scopatz** [comment]
OK, I think whether the tags or some other message is printed to the screen in these events probably is a matter of personal taste and perhaps even what mode you are in (ie maybe you only want it in interactive mode, but not scripts).  I am in favor of having this be customizable with some sensible defaults.

I really do think that we want to have something similar to what iTerm2 has in terms of an interface that comes from the output of commands.  See https://www.iterm2.com/documentation-escape-codes.html This is the most unambiquous case for me.  It is far more likely that a command could print something that looks like a shebang line than something that looks like one of these escapes.  Of course, no matter what we pick there is always this danger when printing out binary files...

I think a PR would be a good place to start.

**Comment by traverseda** [comment]
I've had a chance to do research on this, and have built a prototype around rpyc.

https://gist.github.com/traverseda/52b87330141aaed7300408e80d06512e

This uses rpyc-based "remote objects" which avoids the need to deserialize the object, meaning that third-party programs don't need to deeply integrate with xonsh in order to work.

**Comment by astronouth7303** [comment]
I feel the need to point out that you don't need to do this in xonsh core?

Assuming you're ok with this ui/ux, you could just wrap it in a [macro](https://xon.sh/tutorial_macros.html) like `rpyc!(spam --foo=bar)` and ship it as a xontrib. 

So for your example, something like:

```
🐚 remote = rpyc!(asyncServer.py $PWD)
🐚 for node in remote.iterdir(): print(node.name)
```

I will note that you should probably include the magic incantation to forward sys.stdout and forward logging in a library for applications, to prevent an errant `print()` from corrupting your data stream and to aid in debugging. (I've done this for an application that used msgpack over ssh.)

**Comment by traverseda** [comment]
I have no problem with this being a xontrib, for sure. What I'm running into problems with is the ergonomics of the whole thing. I ultimately want to be able to do stuff like `for node in ls:` or even `for node in $(ssh someremote ls):, making it really easy for users to treat something as both a shell command and a python object, getting things even more tightly integrated between the python and shell.

Of course that really starts to break down when you start, for example, trying to pipe things. Piping `myNewLs | someOtherFunction` obviously won't work under this schema.

Mostly I'm just talking about this here because you're a bunch of smart people with experience writing shells, so I figure you might be able to provide some insight into those deeper problems.

**Comment by astronouth7303** [comment]
Oh, totally agree. I think that `find` should be a builtin in the same way.

I think currently, the same name can be both an alias/command and a function? It's just discouraged for ambiguity reasons.

**Comment by anki-code** [comment]
Now we have SpecModifierAlias (#5443) and XONSH_SUBPROC_OUTPUT_FORMAT (#5377) and  spec.output_format (#5481) and you can return any object e.g. see [xontrib-dalias](https://github.com/anki-code/xontrib-dalias). I'm going to close this issue. Feel free to open new more specific issues.
